core/modules/node/node.pages.inc

```diff
@@ -444,7 +444,21 @@ function node_form_submit($form, &$form_state) {
   if ($node->nid) {
     $form_state['values']['nid'] = $node->nid;
     $form_state['nid'] = $node->nid;
-    $form_state['redirect'] = node_access('view', $node) ? 'node/' . $node->nid : '<front>';
+
+    // Determine whether to redirect user to full page display.
+    $form_state['redirect'] = '<front>';
+    if (node_access('view', $node)) {
+      $form_state['redirect'] = 'node/' . $node->nid;
+
+      // If node type set to prevent direct viewing of full page URL
+      // stay on edit page if user is not permitted.
+      global $user;
+      $pageless = config_get('node.type.' . $node->type, 'settings.pageless');
+      $bypass_pageless = user_access('view page-less content', $user);
+      if ($pageless && !$bypass_pageless) {
+        $form_state['redirect'] = 'node/' . $node->nid . '/edit';
```

Comment:
Maybe a message about this redirection to the edit form can help and prevent confusion. I'm not inspired though.

Reply:
Do you mean a code comment or a message to the user?

Reply:
I assume a message to the user so I took a stab at it.